zerver/lib/actions.py

```diff
@@ -2255,6 +2255,18 @@ def validate_sender_can_write_to_stream(sender: UserProfile,
         else:
             raise JsonableError(_("Only organization administrators can send to this stream."))
 
+    if stream.stream_post_policy == Stream.STREAM_POST_POLICY_RESTRICT_NEW_MEMBERS:
+        if sender.is_realm_admin or is_cross_realm_bot_email(sender.delivery_email):
+            pass
+        elif sender.is_bot and (sender.bot_owner is not None and
+                                sender.bot_owner.is_realm_admin):
+            pass
+        elif sender.is_bot and (sender.bot_owner is not None and
+                                sender.bot_owner.is_new_member):
+            raise JsonableError(_("New members cannot send to this stream."))
+        elif sender.is_new_member:
+            raise JsonableError(_("New members cannot send to this stream."))
```

Comment:
I'm not a huge fan of the duplication with the STREAM_POST_POLICY_ADMINS case.  Maybe the right answer is inverting the order a bit:

```
if sender.is_realm_admin or is_cross_realm_bot_email(...):
    pass
elif sender.is_bot and (... # bot owner case
    pass
elif stream.stream_post_policy == Stream.STREAM_POST_POLICY_ADMINS
    ...
elif stream.stream_post_policy == Stream.STREAM_POST_POLICY_RESTRICT_NEW_MEMBERS
    ...
```